### PR TITLE
`compact_partition` incremental unit test

### PR DESCRIPTION
**Pull request by pfaraone**
## Overview
This PR is a first attempt at a parameterized unit test (using pytest.mark.parametrize) for the `compact_partition` function that covers the incremental compaction of partitions. The test creates a source table with a single partition, stages some arbitrary data into the partition (based on the argument), and then compacts the partition incrementally. The test then validates on the compacted table data

This test is dependent on the local deltacat storage implemented in #175 

## Additional Changes
- `retrying_get` now takes in an optional `short_circuit_on_status` argument to handle early returns and running compaction in a non-EC2 environment (HTTP 403 if not in EC2)
- Optional `deltacat_storage_kwargs` parameter defined for passing in the the connection string when calling the local deltacat storage implementation of the storage APIs

### CODE REVIEW

File: deltacat/compute/compactor/steps/hash_bucket.py [hunk 1]
```diff
@@ -99,6 +100,7 @@ def _group_file_records_by_pk_hash_bucket(
         sort_key_names,
         read_kwargs_provider,
         deltacat_storage,
+        **kwargs,
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/hash_bucket.py:103]: use `deltacat_storage_kwargs` as the `hash_bucket` step itself may require kwargs.

File: deltacat/tests/aws/test_clients.py [hunk 2]
```diff
@@ -48,20 +46,20 @@ def test_sanity(self, requests_mock):
...
-            MockResponse(HTTPStatus.TOO_MANY_REQUESTS, "foo"),
-            MockResponse(HTTPStatus.INTERNAL_SERVER_ERROR, "foo"),
-            MockResponse(HTTPStatus.NOT_IMPLEMENTED, "bar"),
-            MockResponse(HTTPStatus.SERVICE_UNAVAILABLE, "bar"),
-            MockResponse(HTTPStatus.GATEWAY_TIMEOUT, "bar"),
-        ]
-        self.assertEqual(
-            block_until_instance_metadata_service_returns_success().status_code, 200
-        )
+    #     requests_mock.get.side_effect = [
```
Comment (raghumdani) [inline deltacat/tests/aws/test_clients.py]: Why was this commented?

File: deltacat/tests/compactor/test_compaction_session.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+"""
+
+
+def helper_read_s3_contents(
+    s3_resource: ServiceResource, bucket_name: str, key: str
+) -> Dict[str, Any]:
+    response = s3_resource.Object(bucket_name, key).get()
+    file_content: str = response["Body"].read().decode("utf-8")
+    return json.loads(file_content)
+
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:61]: Move these to `test_utils`.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+    def key_name(self) -> str:
+        return self["keyName"]
+
+    @property
+    def key_type(self) -> PartitionKeyType:
+        key_type = self["keyType"]
+        return None if key_type is None else PartitionKeyType(key_type)
+
+
+MAX_RECORDS_PER_FILE: int = 1
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:38]: This can be moved to test constants.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+from enum import Enum
+
+
+class PartitionKeyType(str, Enum):
+    INT = "int"
+    STRING = "string"
+    TIMESTAMP = "timestamp"
+
+
+class PartitionKey(dict):
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:23]: Can we move all these to `test_utils`?

File: deltacat/tests/compactor/test_compaction_session.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    kwargs_for_local_deltacat_storage: Dict[str, Any] = {
+        DATABASE_FILE_PATH_KEY: DATABASE_FILE_PATH_VALUE,
+    }
+    yield kwargs_for_local_deltacat_storage
+    # cleanup the created db file after
+    if os.path.exists(DATABASE_FILE_PATH_VALUE):
+        os.remove(DATABASE_FILE_PATH_VALUE)
+
+
+INCREMENTAL_TEST_CASES = {
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:114]: This can be its own file.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        {"pk_col_1"},  # Primary key columns
+        [],  # Sort key columns
+        [{"key_name": "region_id", "key_type": "int"}],  # Partition keys
+        ["pk_col_1"],  # column_names
+        [pa.array([str(i) for i in range(10)])],  # arrow arrays
+        None,
+        ["1"],
+        pa.Table.from_arrays(
+            [pa.array([str(i) for i in range(10)])], names=["pk_col_1"]
+        ),
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:125]: Can we use named tuple or dictionary for ease of understanding?

File: deltacat/tests/compactor/test_compaction_session.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+        [],
+        ["pk_col_1", "sk_col_1"],
+        [pa.array([str(i) for i in range(10)]), pa.array(["test"] * 10)],
+        None,
+        ["1"],
+        pa.Table.from_arrays(
+            [pa.array([str(i) for i in range(10)]), pa.array(["test"] * 10)],
+            names=["pk_col_1", "sk_col_1"],
+        ),
+        None,
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:144]: You can capture the output to expect in this.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+            PartitionKey.of(
+                partition_key["key_name"], PartitionKeyType(partition_key["key_type"])
+            )
+            for partition_key in in_partition_keys
+        ]
+    # create the source table
+    ds.create_namespace(source_namespace, {}, **ds_mock_kwargs)
+    ds.create_table_version(
+        source_namespace,
+        source_table_name,
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:307]: This can be abstracted to a separated method named `setup_tables`.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+            in_partition_keys,
+            column_names,
+            arrow_arrays,
+            rebase_source_partition_locator,
+            partition_values,
+            expected_result,
+            validation_callback_func,
+        ) in INCREMENTAL_TEST_CASES.items()
+    ],
+    ids=[test_name for test_name in INCREMENTAL_TEST_CASES],
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:245]: I like this.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+    )
+    destination_partition_locator = PartitionLocator.of(
+        destination_table_stream.locator,
+        partition_values,
+        None,
+    )
+    num_workers, worker_instance_cpu = 1, 1
+    total_cpus = num_workers * worker_instance_cpu
+    pgm = PlacementGroupManager(1, total_cpus, worker_instance_cpu).pgs[0]
+    compact_partition_params: Dict[str, Any] = {
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:374]: leverage `CompactPartitionParams.of()` where possible as v2 uses that by default.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        "compaction_artifact_s3_bucket": TEST_S3_RCF_BUCKET_NAME,
+        "compacted_file_content_type": ContentType.PARQUET,
+        "dd_max_parallelism_ratio": 1.0,
+        "deltacat_storage": ds,
+        "destination_partition_locator": destination_partition_locator,
+        "hash_bucket_count": None,
+        "last_stream_position_to_compact": source_partition.stream_position,
+        "list_deltas_kwargs": {**ds_mock_kwargs, **{"equivalent_table_types": []}},
+        "pg_config": pgm,
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:383]: can we parametrize create_placement_group boolean? Its not mandatory to create pg in v2,.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+        "primary_keys": primary_keys,
+        "rebase_source_partition_locator": None,
+        "records_per_compacted_file": MAX_RECORDS_PER_FILE,
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:386]: can we parametrize this as well?

**Review by raghumdani — COMMENTED** [review]
Looks great! Thanks for simplifying tests using pytest parametrize. Please address few minor suggestions.

File: deltacat/tests/aws/test_clients.py (hunk 2 above)
Comment (pfaraone) [reply deltacat/tests/aws/test_clients.py]: Fixed now

File: deltacat/aws/clients.py [hunk 14]
```diff
@@ -121,11 +121,13 @@ def block_until_instance_metadata_service_returns_success(
...
     https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
     """
+    if _get_url(url).status_code == HTTPStatus.FORBIDDEN:
+        return None
     return retrying_get(url, retry_strategy, wait_strategy, stop_strategy)
 
 
 def _get_session_from_kwargs(input_kwargs):
-    block_until_instance_metadata_service_returns_success()
+    # block_until_instance_metadata_service_returns_success()
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/aws/clients.py]: Fixed now

File: deltacat/compute/compactor/compaction_session.py [hunk 15]
```diff
@@ -173,7 +178,7 @@ def compact_partition(
         round_completion_file_s3_url = None
         if partition:
             logger.info(f"Committing compacted partition to: {partition.locator}")
-            partition = deltacat_storage.commit_partition(partition)
+            partition = deltacat_storage.commit_partition(partition, **kwargs)
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 11 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py:409]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 12 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py:419]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 9 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py:356]: This get's kind of tricky with parametrize as we are using some of the parameters to construct the tables. Will need to see how best to accomplish this - using `indirect` may work - https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments

File: deltacat/tests/compactor/test_compaction_session.py (hunk 11 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Done

File: deltacat/tests/local_deltacat_storage/__init__.py [hunk 16]
```diff
@@ -188,9 +188,9 @@ def list_deltas(
 
     for delta in all_deltas:
         if (
-            not first_stream_position or first_stream_position <= delta.stream_position
```
Comment (Zyiqin-Miranda) [inline deltacat/tests/local_deltacat_storage/__init__.py:191]: Do we change here from [first_stream_position, last_stream_position) to (first_stream_position, last_stream_position]? What's the context of this change?

**Review by Zyiqin-Miranda — COMMENTED** [review]
LGTM! Thanks Pierce for adding all these useful tests

File: deltacat/tests/compactor/test_compaction_session.py (hunk 6 above)
Comment (raghumdani) [reply deltacat/tests/compactor/test_compaction_session.py:116]: Many comments on this file are not yet addressed.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        ),
+        None,
+        None,
+        True,
+        False,
+        True,
+        MAX_RECORDS_PER_FILE,
+        None,
+    ),
+    "4-incremental-duplicate-pk": (
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:195]: Please add few more test cases for duplicate pk with decimal columns, and int columns and multi pk as well.

File: deltacat/tests/compactor/test_compaction_session.py [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+            cleanup_prev_table,
+            use_prev_compacted,
+            create_placement_group,
+            records_per_compacted_file,
+            hash_bucket_count,
+        ) in INCREMENTAL_TEST_CASES.items()
+    ],
+    ids=[test_name for test_name in INCREMENTAL_TEST_CASES],
+)
+def test_compact_partition_incremental(
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session.py:291]: Is testing, rebase and then incremental part of separate PR? I think that is a critical case which must be tested to ensure compactor is logically doing it's job.,

**Review by raghumdani — COMMENTED** [review]
Looks good overall. Please address minor comments.

File: deltacat/compute/compactor/compaction_session.py [hunk 19]
```diff
@@ -118,6 +126,7 @@ def compact_partition(
     object_store: Optional[IObjectStore] = RayPlasmaObjectStore(),
     s3_client_kwargs: Optional[Dict[str, Any]] = None,
     deltacat_storage=unimplemented_deltacat_storage,
+    deltacat_storage_kwargs: Optional[Dict[str, Any]] = None,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py:129]: Done

File: deltacat/compute/metastats/meta_stats.py [hunk 20]
```diff
@@ -118,6 +118,7 @@ def collect_from_partition(
     stat_results_s3_bucket: Optional[str] = None,
     metastats_results_s3_bucket: Optional[str] = None,
     deltacat_storage=unimplemented_deltacat_storage,
+    deltacat_storage_kwargs: Optional[Dict[str, Any]] = None,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/metastats/meta_stats.py:121]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 9 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Done @raghumdani

File: deltacat/compute/compactor/compaction_session.py [hunk 21]
```diff
@@ -162,7 +171,8 @@ def compact_partition(
             object_store,
             s3_client_kwargs,
             deltacat_storage,
-            **kwargs,
+            deltacat_storage_kwargs,
+            **list_deltas_kwargs,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py:176]: Fixed - the only change with the discover_deltas signature is now that it takes in a deltacat_storage_kwargs args in addition to the kwargs (in which we are passing **list_deltas_kwargs)

File: deltacat/compute/compactor/compaction_session.py [hunk 22]
```diff
@@ -330,7 +345,7 @@ def _execute_compaction_round(
         rebase_source_partition_locator,
         rebase_source_partition_high_watermark,
         deltacat_storage,
-        **list_deltas_kwargs,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py:333]: Fixed

File: deltacat/compute/compactor/compaction_session.py [hunk 23]
```diff
@@ -502,6 +523,7 @@ def _execute_compaction_round(
         enable_profiler=enable_profiler,
         metrics_config=metrics_config,
         object_store=object_store,
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py:528]: Unintended. Reverting to not passing kwargs as per main

File: deltacat/compute/compactor/compaction_session.py [hunk 24]
```diff
@@ -585,6 +604,8 @@ def _execute_compaction_round(
         s3_table_writer_kwargs=s3_table_writer_kwargs,
         object_store=object_store,
         deltacat_storage=deltacat_storage,
+        deltacat_storage_kwargs=deltacat_storage_kwargs,
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/compute/compactor/compaction_session.py:609]: Unintended. Reverting to not passing kwargs as per main

File: deltacat/tests/compactor/test_compaction_session.py (hunk 18 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Yes - rebase and then incremental will be in a different PR

File: deltacat/tests/compactor/testcases.py [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+        ),
+        None,
+        None,
+        True,
+        False,
+        True,
+        MAX_RECORDS_PER_FILE,
+        None,
+    ),
+    "5-incremental-decimal-pk": (
```
Comment (raghumdani) [inline deltacat/tests/compactor/testcases.py:123]: Insert few duplicates in here.

File: deltacat/tests/compactor/testcases.py [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+        ),
+        None,
+        None,
+        True,
+        False,
+        True,
+        MAX_RECORDS_PER_FILE,
+        None,
+    ),
+    "7-incremental-integer-pk": (
```
Comment (raghumdani) [inline deltacat/tests/compactor/testcases.py:153]: Same has here.

File: deltacat/tests/local_deltacat_storage/__init__.py (hunk 16 above)
Comment (raghumdani) [reply deltacat/tests/local_deltacat_storage/__init__.py:191]: It was bug.

File: deltacat/tests/compactor/test_compaction_session_incremental.py [hunk 27]
```diff
@@ -0,0 +1,341 @@
...
+    records_per_compacted_file_param,
+    hash_bucket_count_param,
+):
+
+    """
+    TODO Test Cases:
+    1. incremental w/wout round completion file
+    2. Backfill w/wout round completion
+    3. Rebase w/wout round completion file
+    4. Rebase then incremental (use same round completion file)
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session_incremental.py:246]: As far as this file is concerned, it is complete. We can move this TODO to testcases.py.

File: deltacat/tests/compactor/test_compaction_session_incremental.py [hunk 28]
```diff
@@ -0,0 +1,341 @@
...
+    assert compacted_table.equals(expected_result)
+    if (
+        validation_callback_func is not None
+        and validation_callback_func_kwargs is not None
+    ):
+        validation_callback_func(**validation_callback_func_kwargs)
+    # if not cleanup_prev_table:
+    #     pass
+    # else:
+    #     request.getfixturevalue("cleanup_database_between_executions")
```
Comment (raghumdani) [inline deltacat/tests/compactor/test_compaction_session_incremental.py:341]: Avoid dead code.

**Review by raghumdani — COMMENTED** [review]
Looks good overall. Please address few more minor comments.

File: deltacat/tests/compactor/testcases.py (hunk 25 above)
Comment (pfaraone) [reply deltacat/tests/compactor/testcases.py]: Added additional test cases instead of modifying this specific test case

File: deltacat/tests/compactor/testcases.py [hunk 29]
```diff
@@ -0,0 +1,307 @@
...
+        ),
+        None,
+        None,
+        True,
+        False,
+        True,
+        MAX_RECORDS_PER_FILE,
+        None,
+    ),
+    "10-incremental-decimal-pk-multi-dup": (
```
Comment (raghumdani) [inline deltacat/tests/compactor/testcases.py:275]: Perfect!

File: deltacat/tests/compactor/testcases.py [hunk 30]
```diff
@@ -0,0 +1,307 @@
...
+        [
+            {
+                "key_name": "sk_col_1",
+            },
+        ],
+        [],
+        ["pk_col_1", "sk_col_1"],
+        [
+            pa.array([0.1] * 4 + [0.2] * 4 + [0.3] * 4 + [0.4] * 4 + [0.5] * 4),
+            pa.array([i for i in range(20)]),
```
Comment (raghumdani) [inline deltacat/tests/compactor/testcases.py:288]: Reverse the order so we know sorting works.

**Review by raghumdani — APPROVED** [review]
LGTM!

File: deltacat/tests/compactor/testcases.py (hunk 30 above)
Comment (pfaraone) [reply deltacat/tests/compactor/testcases.py]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 7 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Will pick this up in subsequent PR

File: deltacat/tests/compactor/test_compaction_session.py (hunk 6 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Done

File: deltacat/tests/compactor/test_compaction_session.py (hunk 4 above)
Comment (pfaraone) [reply deltacat/tests/compactor/test_compaction_session.py]: Done - moved to within `common.py` within the same package
